### PR TITLE
Implement getstrbyname/MVM_unicode_string_from_name to retrieve grapheme clusters by name

### PR DESCRIPTION
**Pull request by samcv**
Returns a string from its Unicode name. Unlike MVM_unicode_lookup_by_name, this works
for multiple codepoints. This allows us to support Emoji sequences and other potential
names which are sequences of codepoints.

Working example: ```nqp-m -e "say(nqp::getstrfromname('person golfing: medium-light skin tone'))" #> 🏌🏼```

### CODE REVIEW

File: src/core/oplist [hunk 1]
```diff
@@ -800,6 +800,7 @@ indexingoptimized   w(str) r(str) :pure
 captureinnerlex     r(obj) :noinline
 unicmp_s            w(int64) r(str) r(str) r(int64) r(int64) r(int64) :pure
 setdispatcherfor    r(obj) r(obj)
+getstrbyname         w(str) r(str) :pure
```
Comment (jnthn) [inline src/core/oplist]: The `w(str)` seems a space out of line with all those above it.

File: src/strings/unicode_ops.c [hunk 2]
```diff
@@ -108,13 +108,41 @@ MVMGrapheme32 MVM_unicode_lookup_by_name(MVMThreadContext *tc, MVMString *name)
...
+
     if (!codepoints_by_name) {
         generate_codepoints_by_name(tc);
     }
     HASH_FIND(hash_handle, codepoints_by_name, cname, strlen((const char *)cname), result);
     MVM_free(cname);
+
     return result ? result->codepoint : -1;
 }
+/* XXX move this so it's included properly */
```
Comment (jnthn) [inline src/strings/unicode_ops.c]: Either should be done, or shouldn't but the comment not left behind.

File: src/strings/unicode_ops.c [hunk 3]
```diff
@@ -108,13 +108,41 @@ MVMGrapheme32 MVM_unicode_lookup_by_name(MVMThreadContext *tc, MVMString *name)
...
+MVMString * MVM_string_from_grapheme(MVMThreadContext *tc, MVMCodepoint cp) {
```
Comment (jnthn) [inline src/strings/unicode_ops.c]: This is called `string_from_grapheme`, but actually is taking a codepoint, which is not always a grapheme.

File: src/strings/unicode_ops.c [hunk 4]
```diff
@@ -108,13 +108,41 @@ MVMGrapheme32 MVM_unicode_lookup_by_name(MVMThreadContext *tc, MVMString *name)
...
+    MVMString *s;
+    MVMGrapheme32 g;
+    MVMNormalizer norm;
+
+    MVM_unicode_normalizer_init(tc, &norm, MVM_NORMALIZE_NFG);
+    if (!MVM_unicode_normalizer_process_codepoint_to_grapheme(tc, &norm, cp, &g)) {
+        MVM_unicode_normalizer_eof(tc, &norm);
+        g = MVM_unicode_normalizer_get_grapheme(tc, &norm);
```
Comment (jnthn) [inline src/strings/unicode_ops.c]: This code makes the (so far as I can tell, correct in Unicode 9) assumption that a single codepoint will always result in a single grapheme under NFG. I'm just a little nervous about this assumption as it's not true for NFC. I guess it's unlikely that such a case will arise in the future too. But it caught my eye, and thought it was worth mentioning.

File: src/strings/unicode.h [hunk 5]
```diff
@@ -16,6 +16,11 @@ struct MVMUnicodeNameRegistry {
     MVMGrapheme32 codepoint;
     UT_hash_handle hash_handle;
 };
+struct MVMUnicodeGraphemeNameRegistry {
+    char *name;
+    MVMint32 structitem;
```
Comment (jnthn) [inline src/strings/unicode.h]: Not entirely keen on `structitem` as a name...it seems more of an index to the codepoint sequence?

File: src/strings/unicode_ops.c [hunk 6]
```diff
@@ -400,3 +440,42 @@ void MVM_unicode_release(MVMThreadContext *tc)
...
+    if (!result)
+        return tc->instance->str_consts.empty;
+
+    uni_seq = uni_seq_enum[result->structitem];
+    /* The first element is the number of codepoints in the sequence */
+    array_size = uni_seq[0];
+    MVM_unicode_normalizer_init(tc, &norm, MVM_NORMALIZE_NFG);
+    for (int i = 1; i < array_size; i++) {
+        cp = uni_seq[i];
+        MVM_unicode_normalizer_process_codepoint_full(tc, &norm, cp, &cp);
```
Comment (jnthn) [inline src/strings/unicode_ops.c]: This discards the return value. I guess at the moment we assume that all sequences are going to produce a single grapheme. That's reasonable, but we should at the very least stick in an `assert` that we really get `0` back from this.

Comment (jnthn) [inline src/strings/unicode_ops.c]: Also, re-using `cp` for the out parameter is a bit confusing.

File: src/strings/unicode_ops.c [hunk 7]
```diff
@@ -400,3 +440,42 @@ void MVM_unicode_release(MVMThreadContext *tc)
...
+    MVM_free(cname);
+    /* If we can't find a result return an empty string */
+    if (!result)
+        return tc->instance->str_consts.empty;
+
+    uni_seq = uni_seq_enum[result->structitem];
+    /* The first element is the number of codepoints in the sequence */
+    array_size = uni_seq[0];
+    MVM_unicode_normalizer_init(tc, &norm, MVM_NORMALIZE_NFG);
+    for (int i = 1; i < array_size; i++) {
```
Comment (jnthn) [inline src/strings/unicode_ops.c]: If the first element is the number of codepoints in the sequence, but then we start at `1`, will we not miss the last codepoint?

File: src/strings/unicode_ops.c [hunk 8]
```diff
@@ -400,3 +440,42 @@ void MVM_unicode_release(MVMThreadContext *tc)
...
+        cp = uni_seq[i];
+        MVM_unicode_normalizer_process_codepoint_full(tc, &norm, cp, &cp);
+    }
+    MVM_unicode_normalizer_process_codepoint_norm_terminator(tc, &norm, cp, &result_graph);
```
Comment (jnthn) [inline src/strings/unicode_ops.c]: I don't understand `cp` being fed into the normalizer here; either it's what we already put in, or it's whatever was already written into `cp` by the normalizer the last time around the loop. Should this just be calling the `eof`` function?

File: src/strings/unicode_ops.c [hunk 9]
```diff
@@ -400,3 +440,42 @@ void MVM_unicode_release(MVMThreadContext *tc)
...
+    MVM_unicode_normalizer_cleanup(tc, &norm);
+
+    return MVM_string_from_grapheme(tc, result_graph);
```
Comment (jnthn) [inline src/strings/unicode_ops.c]: By this point we have an NFG grapheme. But then `MVM_string_from_grapheme` is running it through the normalizer again. That's a bit wasteful, but also the normalizer was designed to be fed codepoints, not graphemes, so there's a risk of oddness going on from that.

File: src/strings/unicode_ops.c (hunk 4 above)
Comment (jnthn) [inline src/strings/unicode_ops.c]: Though as mentioned below, this whole pass through the normalizer may not make sense.

File: src/strings/unicode_ops.c (hunk 2 above)
Comment (jnthn) [inline src/strings/unicode_ops.c]: fwiw, I think this particular routine would be better in `src/string/ops.c` and with a decl in `src/string/ops.h`.

File: src/strings/unicode_ops.c (hunk 8 above)
Comment (samcv) [inline src/strings/unicode_ops.c]: Yes I think it should. Will make that change.

File: src/strings/unicode.h (hunk 5 above)
Comment (samcv) [inline src/strings/unicode.h]: Yeah it's the index. Name changed.

File: src/core/oplist [hunk 10]
```diff
@@ -800,6 +800,7 @@ indexingoptimized   w(str) r(str) :pure
 captureinnerlex     r(obj) :noinline
 unicmp_s            w(int64) r(str) r(str) r(int64) r(int64) r(int64) :pure
 setdispatcherfor    r(obj) r(obj)
+getstrfromname        w(str) r(str) :pure
```
Comment (jnthn) [inline src/core/oplist]: Whitespace is a bit off again.

File: src/strings/unicode_ops.c [hunk 11]
```diff
@@ -400,3 +412,35 @@ void MVM_unicode_release(MVMThreadContext *tc)
...
+    MVMGrapheme32 result_graph;
+    const MVMint32 * uni_seq;
+    int array_size;
+    MVMint32 cp;
+    MVMNormalizer norm;
+
+    result_graph = MVM_unicode_lookup_by_name(tc, name);
+    /* If it's just a codepoint, return that */
+    if (result_graph >= 0)
+        return MVM_string_from_grapheme(tc, result_graph);
```
Comment (jnthn) [inline src/strings/unicode_ops.c]: We could eliminate `MVM_string_from_grapheme` and just use `MVM_string_chr` here - which will take care of singleton decomposition, which this doesn't.

**Comment by samcv** [comment]
Should have made all the changes you requested.

File: src/strings/unicode_ops.c [hunk 12]
```diff
@@ -400,3 +412,35 @@ void MVM_unicode_release(MVMThreadContext *tc)
...
+MVMString * MVM_unicode_string_from_name(MVMThreadContext *tc, MVMString *name) {
+    MVMuint64 size;
+    char *cname = MVM_string_ascii_encode(tc, name, &size, 0);
+    MVMUnicodeGraphemeNameRegistry *result;
+
+    MVMGrapheme32 result_graph;
+    const MVMint32 * uni_seq;
+    int array_size;
+    MVMint32 cp;
+    MVMNormalizer norm;
```
Comment (jnthn) [inline src/strings/unicode_ops.c]: The variables `cp` and `norm` are unused.

File: src/strings/unicode_ops.c [hunk 13]
```diff
@@ -400,3 +412,33 @@ void MVM_unicode_release(MVMThreadContext *tc)
...
+    MVMUnicodeGraphemeNameRegistry *result;
+
+    MVMGrapheme32 result_graph;
+    const MVMint32 * uni_seq;
+    int array_size;
+
+    result_graph = MVM_unicode_lookup_by_name(tc, name);
+    /* If it's just a codepoint, return that */
+    if (result_graph >= 0)
+        return MVM_string_chr(tc, result_graph);
```
Comment (jnthn) [inline src/strings/unicode_ops.c]: It seems if we take this path, then we leak `cname` (because we never hit the `MVM_free` below).

Comment (samcv) [inline src/strings/unicode_ops.c]: Fixed.

**Pull request merged**